RecoEgamma/EgammaTools/plugins/HGCalElectronIDValueMapProducer.cc

```diff
@@ -0,0 +1,276 @@
+// -*- C++ -*-
+//
+// Package:   RecoEgamma/EgammaTools
+// Class:    HGCalElectronIDValueMapProducer
+//
+/**\class HGCalElectronIDValueMapProducer HGCalElectronIDValueMapProducer.cc RecoEgamma/EgammaTools/plugins/HGCalElectronIDValueMapProducer.cc
+
+ Description: [one line class summary]
+
+ Implementation:
+    [Notes on implementation]
+*/
+//
+// Original Author:  Nicholas Charles Smith
+//      Created:  Wed, 05 Apr 2017 12:17:43 GMT
+//
+//
+
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/Common/interface/ValueMap.h"
+
+#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
+#include "DataFormats/EgammaCandidates/interface/GsfElectronFwd.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+
+#include "RecoEgamma/EgammaTools/interface/HGCalEgammaIDHelper.h"
+#include "RecoEgamma/EgammaTools/interface/LongDeps.h"
+
+class HGCalElectronIDValueMapProducer : public edm::stream::EDProducer<> {
+  public:
+    explicit HGCalElectronIDValueMapProducer(const edm::ParameterSet&);
+    ~HGCalElectronIDValueMapProducer() override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  private:
+    void beginStream(edm::StreamID) override;
+    void produce(edm::Event&, const edm::EventSetup&) override;
+    void endStream() override;
+
+    // ----------member data ---------------------------
+    edm::EDGetTokenT<edm::View<reco::GsfElectron>> electronsToken_;
+    float radius_;
+
+    static const std::vector<std::string> valuesProduced_;
+    std::map<const std::string, std::vector<float>> maps_;
+
+    std::unique_ptr<HGCalEgammaIDHelper> eIDHelper_;
+};
+
+// All the ValueMap names to output are defined in the auto-generated python cfi
+// so that potential consumers can configure themselves in a simple manner
+// Would be cool to use compile-time validation, but need constexpr strings, e.g. std::string_view in C++17
+const std::vector<std::string> HGCalElectronIDValueMapProducer::valuesProduced_ = {
```

Comment:
hi @nsmith- could you define instead a list of enums with a enum->string lookup std::vector? That would remove lots of string comparisons.

Reply:
I did ask for the same during the review.

However,
the effort for doing this is not really worth it
-  in the longer term perspective of this code the map producer is supposed to be eventually replaced by something that writes a struct or a class once the variables are settled
- the CPU gain from this effort is apparently about 0.1% (0.05/54) of CPU taken by this module https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_10_0_0_pre1-sign978.pat-from-aod.2023PU200.20234.2.pp/182

Reply:
p.s. how did you get igprof to give a valid sqlite file?